File: ext/c++/common/component/logic/center.cpp
/*
 * qianqians
 * 2016-7-9
 * center.cpp
 */
#include <boost/uuid/uuid.hpp>
#include <boost/uuid/uuid_generators.hpp>
#include <boost/uuid/uuid_io.hpp>
#include <boost/lexical_cast.hpp>
#include <boost/thread.hpp>

#include <acceptnetworkservice.h>
#include <process_.h>
#include <Ichannel.h>
#include <Imodule.h>
#include <channel.h>
#include <juggleservice.h>
#include <timerservice.h>

#include <config.h>

#include <module/centermodule.h>
#include <module/gm_centermodule.h>
#include <module/logic_call_centermodule.h>

#include "svrmanager.h"
#include "logicsvrmanager.h"
#include "logic_svr_msg_handle.h"
#include "svr_msg_handle.h"
#include "gm_msg_handle.h"

void main(int argc, char * argv[]) {
	auto svr_uuid = boost::lexical_cast<std::string>(boost::uuids::random_generator()());

	if (argc <= 1) {
		std::cout << "non input start argv" << std::endl;
		return;
	}

	std::string config_file_path = argv[1];
	auto _config = std::make_shared<config::config>(config_file_path);
	if (argc >= 3) {
		_config = _config->get_value_dict(argv[2]);
	}

	std::shared_ptr<juggle::process> _svr_process = std::make_shared<juggle::process>();
	
	std::shared_ptr<center::svrmanager> _svrmanager = std::make_shared<center::svrmanager>();
	std::shared_ptr<center::logicsvrmanager> _logicsvrmanager = std::make_shared<center::logicsvrmanager>();
	std::shared_ptr<module::center> _center = std::make_shared<module::center>();
	_center->sigreg_serverhandle.connect(std::bind(&reg_server, _svrmanager, _logicsvrmanager, std::placeholders::_1, std::placeholders::_2, std::placeholders::_3, std::placeholders::_4));
	_svr_process->reg_module(_center);
	
	std::shared_ptr<module::logic_call_center> _logic_call_center = std::make_shared<module::logic_call_center>();
	_logic_call_center->sigreq_get_server_addresshandle.connect(std::bind(&req_get_server_address, _logicsvrmanager, _svrmanager, std::placeholders::_1, std::placeholders::_2));
	_svr_process->reg_module(_logic_call_center);

	auto host_ip = _config->get_value_string("ip");
	auto host_port = (short)_config->get_value_int("port");
	auto _host_service = std::make_shared<service::acceptnetworkservice>(host_ip, host_port, _svr_process);

	std::shared_ptr<center::gmmanager> _gmmanager = std::make_shared<center::gmmanager>();
	std::shared_ptr<juggle::process> _gm_process = std::make_shared<juggle::process>();
	std::shared_ptr<module::gm_center> _gm_center = std::make_shared<module::gm_center>();
	_gm_center->sigconfirm_gmhandle.connect(std::bind(&confirm_gm, _gmmanager, std::placeholders::_1));
	_gm_center->sigclose_clutterhandle.connect(std::bind(&close_clutter, _gmmanager, _svrmanager, std::placeholders::_1));
	_gm_process->reg_module(_gm_center);

	auto gm_ip = _config->get_value_string("gm_ip");
	auto gm_port = (short)_config->get_value_int("gm_port");
	auto _gm_service = std::make_shared<service::acceptnetworkservice>(gm_ip, gm_port, _gm_process);

	std::shared_ptr<service::juggleservice> _juggleservice = std::make_shared<service::juggleservice>();
	_juggleservice->add_process(_svr_process);
	_juggleservice->add_process(_gm_process);

	int64_t tick = clock();
	int64_t tickcount = 0;

	std::shared_ptr<service::timerservice> _timerservice = std::make_shared<service::timerservice>(tick);

	while (true)
	{
		int64_t tmptick = (clock() & 0xffffffff);
		if (tmptick < tick)
		{
			tickcount += 1;
			tmptick = tmptick + tickcount * 0xffffffff;
		}
		tick = tmptick;

		try {
			_host_service->poll(tick);
			_gm_service->poll(tick);
			_juggleservice->poll(tick);
			_timerservice->poll(tick);
		}
		catch (std::exception e) {
			std::cout << "system exception:" << e.what() << std::endl;
			break;
		}

		tmptick = (clock() & 0xffffffff);
		if (tmptick < tick)
		{
			tickcount += 1;
			tmptick = tmptick + tickcount * 0xffffffff;
		}
		int64_t ticktime = (tmptick - tick);
		tick = tmptick;

		if (ticktime < 50) {
			boost::thread::sleep(boost::get_system_time() + boost::posix_time::microseconds(15));
		}
	}
}

File: component/center/center.cs
using System;
using System.Threading;

namespace center
{
	class center
	{
		public center(string[] args)
		{
			uuid = System.Guid.NewGuid().ToString();

			config.config _config = new config.config(args[0]);
			if (args.Length > 1)
			{
				_config = _config.get_value_dict(args[1]);
			}

            closeHandle = new closehandle();

            var log_level = _config.get_value_string("log_level");
            if (log_level == "debug")
            {
                log.log.logMode = log.log.enLogMode.Debug;
            }
            else if (log_level == "release")
            {
                log.log.logMode = log.log.enLogMode.Release;
            }
            var log_file = _config.get_value_string("log_file");
            log.log.logFile = log_file;
            var log_dir = _config.get_value_string("log_dir");
            log.log.logPath = log_dir;
            {
                if (!System.IO.Directory.Exists(log_dir))
                {
                    System.IO.Directory.CreateDirectory(log_dir);
                }
            }

            _svrmanager = new svrmanager ();
			_hubmanager = new hubmanager ();

            var _clutter = new clutter();

			juggle.process _center_process = new juggle.process ();

			_svr_call_center = new module.center ();
			_svr_msg_handle = new svr_msg_handle (_svrmanager, _hubmanager, _clutter);
			_svr_call_center.onreg_server += _svr_msg_handle.reg_server;
			_center_process.reg_module (_svr_call_center);

            var _hub_call_center = new module.hub_call_center();
            var _hub_msg_handle = new hub_msg_handle(_svrmanager, _hubmanager, _clutter);
            _hub_call_center.onclosed += _hub_msg_handle.closed;
            _center_process.reg_module(_hub_call_center);

            var ip = _config.get_value_string("ip");
			var port = (short)_config.get_value_int("port");
			_accept_svr_service = new service.acceptnetworkservice (ip, port, _center_process);

			_gmmanager = new gmmanager ();

			juggle.process _gm_process = new juggle.process ();

			_gm_center = new module.gm_center ();
			_gm_msg_handle = new gm_msg_handle(_gmmanager, _svrmanager, _hubmanager, _clutter);
			_gm_center.onconfirm_gm += _gm_msg_handle.confirm_gm;
			_gm_center.onclose_clutter += _gm_msg_handle.close_clutter;
            _gm_center.onclose_zone += _gm_msg_handle.close_zone;
            _gm_center.onreload += _gm_msg_handle.reload;
            _gm_process.reg_module(_gm_center);

			var gm_ip = _config.get_value_string("gm_ip");
			var gm_port = (short)_config.get_value_int("gm_port");
			_accept_gm_service = new service.acceptnetworkservice (gm_ip, gm_port, _gm_process);

			timer = new service.timerservice ();

			_juggle_service = new service.juggleservice ();
			_juggle_service.add_process (_center_process);
			_juggle_service.add_process (_gm_process);
		}

		public Int64 poll()
        {
            Int64 tick_begin = timer.poll();

            try
            {
                _juggle_service.poll(tick_begin);
            }
            catch (juggle.Exception e)
            {
                log.log.error(new System.Diagnostics.StackFrame(true), tick_begin, e.Message);
            }
            catch (System.Exception e)
            {
                log.log.error(new System.Diagnostics.StackFrame(true), tick_begin, "{0}", e);
            }

            System.GC.Collect();

            Int64 tick_end = timer.refresh();

            return tick_end - tick_begin;
        }

		private static void Main (string[] args)
		{
			if (args.Length <= 0)
            {
                log.log.error(new System.Diagnostics.StackFrame(true), service.timerservice.Tick, "non input start argv");
				return;
			}

			center _center = new center(args);

            while (true)
			{
                var tmp = _center.poll();

                if (closeHandle.is_close)
                {
                    System.Threading.Thread.Sleep(100);
                    return;
                }

                if (tmp < 50)
                {
                    System.Threading.Thread.Sleep(15);
                }
			}
		}

		private String uuid;

        public static closehandle closeHandle;

        private service.acceptnetworkservice _accept_svr_service;

		private module.center _svr_call_center;
		private svr_msg_handle _svr_msg_handle;
		private svrmanager _svrmanager;

		private hubmanager _hubmanager;

		private service.acceptnetworkservice _accept_gm_service;
		private module.gm_center _gm_center;
		private gm_msg_handle _gm_msg_handle;
		private gmmanager _gmmanager;

		private service.juggleservice _juggle_service;
		private service.timerservice timer;
	}
}
